Make `make quick` quick

tomleb opened the pull request:
## Issue

`make quick` is pretty slow when developing Rancher. I don't use it for that reason, instead I use `make build` because it's much quicker (because dapper currently caches go build better than our dockerfiles).

This PR attempt to make `make quick` much faster by leveraging "more modern" docker build features: [RUN --mount=type=cache](https://docs.docker.com/reference/dockerfile/#run---mounttypecache).

First, the result..

### Results

**TL;DR**

|Step|Before changes|After changes|Speed up|
|:---  |    ----:|   ----:|   ----:|
|Cold build|~5m7s|~3m18s|1.55x|
|No changes|~7s|~7s|1x|
|Editing Go file|~3m21s|~29s|6.9x|
|Updating go.mod|~4m54s|~49s|6.0x|

<details>
<summary>Gimme `time` output</summary>



Here are the results from two scenarios I typically run into: (1) editing Go files directly in Rancher and (2) dependency changes in go.mod (either due to manual upgrades or git rebase). 

**Cold build**

Without PR:

```
real    5m7.062s
user    0m2.030s
sys     0m1.131s
```

With PR: Faster because go mod download was cached for the agent image so we didn't need to download those deps twice.

```
real    3m18.150s
user    0m2.028s
sys     0m1.153s
```

**No changes**

I just re-run `make quick` right after another `make quick` with no changes. Not very practical test but shows us the minimum amount of time a build will take.

Without PR:

```
real    0m6.233s
user    0m0.774s
sys     0m0.495s
```

With PR:

```
real    0m5.133s
user    0m0.845s
sys     0m0.461s
```

**Editing Go file**

This scenario is pretty simple. I'll edit an error message in file `pkg/wrangler/context.go` and compare how long that took to build. Each command simply runs `time make quick`.

Without PR:

```
real    3m21.571s
user    0m1.040s
sys     0m0.608s
```

With PR:

```
real    0m27.922s
user    0m0.814s
sys     0m0.426s
```

**Updating go.mod**

This scenario changes the Steve import from `v0.6.1` to `e1061a86cd67b47ca8a0a6b23c12dc6ce1a952a5`

Without PR:

```
real    4m54.650s
user    0m1.414s
sys     0m0.832s
```

With PR:

```
real    0m47.342s
user    0m0.913s
sys     0m0.464s
```

</details>

## Explanation

Using `--mount=type=cache,target=/root/.cache` in `RUN` directives helps us in two ways.

For `go mod download` commands, even if that layer needs to be re-run (due to a dependency change), the Go modules will most often be cached in that cache directory (the PR sets `GOMODCACHE=/root/.cache/go/modcache`) which means it doesn't need to download ALL the dependencies if only a single one changes. (Also note that both `Dockerfile` and `Dockerfile.agent` download go deps, which with these changes only need to be downloaded once now)

For `go build` commands, the compiler caches build artifacts to speed up compilation. Our current Dockerfile never keeps this information so the compiler ends up having to re-build everything even if a single file changes. With the changes, we're caching this in `GOCACHE` (`/root/.cache/go/cache`), so now compilation should be much shorter. (Think how much dependencies we have that Go has to go through, which rarely changes when doing the usual code-compile-test loop).

## The cache

The caches from `--mount=type=cache` are not used when docker build is provided with `--no-cache`. I believe we want this for SLSA reasons (cc: @pjbgf), but for developing I think it's a must to have the improved caching.

**Note:** Don't forget to run `docker buildx prune` once in a while to delete unused cached.

## Future improvement

Currently, building the agent image depends on the whole Rancher image being built. However, we see that we only need a subset of the rancher image:

```
COPY --from=rancher /var/lib/rancher-data /var/lib/rancher-data
COPY --from=rancher /usr/bin/tini /usr/bin/
```

We should be able to build only the agent image when needed instead of going through both. (Of 30s build, rancher takes ~20s and the agent ~10s, so for building the agent image we should be able to shave off another 20s or so I think..)

Here's my attempt at this: https://github.com/rancher/rancher/pull/49636. 

---

cc: @pmatseykanets I'd be curious what kind of numbers you get with this, I remember your compile time to be pretty slow.

pjbgf reviewed (approved):
Nice one, @tomleb! Overall LGTM.

I linked an internal issue to track the changes that will be required for Prime as a follow-up.

pjbgf commented:
> The caches from --mount=type=cache are not used when docker build is provided with --no-cache. I believe we want this for SLSA reasons (cc: @pjbgf), but for developing I think it's a must to have the improved caching.

@tomleb +1 on caching for test/dev and its measurable impact on productivity.
For releases, this isn't a hard-requirement from SLSA, however from a security stand-point we are better off starting from a fresh pull of dependencies to avoid potential cache poisoning attacks.

tomleb commented:
I'm investigating some potential build time regression with this on mac. Doing some more testing with some folks (I don't have a mac) and worse case we'll disable this cache when for mac users.

tomleb commented:
Just an update here: I removed `--link` because after extensive testing I couldn't find a performance boost. I added an id to the mount because it makes it easier to find with `docker buildx du --verbose --filter type=exec.cachemount` to troubleshoot caches.

Still troubleshooting the mac case.

mallardduck commented:
FWIW, this seems to provide some pretty positive benefits on my mac. I'm using a mac studio though so it's pretty beefy specs that may not be reflective of all arm macs.

tomleb commented:
Peter tried running it again after a pretty aggressive prune of its docker (docker desktop?) builder and it seems like it's now building faster. 

I'll rebase and afterwards I think we're good with the changes.

pmatseykanets reviewed (approved):
👍
